### PR TITLE
receiverID and header input and output for all messages

### PR DESCRIPTION
**Pull request by Angeleon**
As receiverID is desired for CARLA control, here is a suggestion for the first step on how to implement it.

### CODE REVIEW

File: src/traj.c [hunk 1]
```diff
@@ -27,6 +29,7 @@ static DebugStrings_t TRAJInfoDescription = 		{"Trajectory info",	"",	&printU8};
  *		EMSGSIZE	if trajectory name is too long
  */
 ssize_t encodeTRAJMessageHeader(
+	const uint32_t receiverID,  const uint8_t messageCounter,
```
Comment (ceciliahernqvist) [inline src/traj.c:32]: Small detail but move messageCounter to its own row to follow the structure of one parameter per row.

File: iso22133.c [hunk 2]
```diff
@@ -863,7 +863,7 @@ ssize_t encodeHEABMessage(const struct timeval *heabTime, const enum ControlCent
 	}
 
 	// Construct header
-	HEABData.header = buildISOHeader(MESSAGE_ID_HEAB, sizeof (HEABData), debug);
+	HEABData.header = buildISOHeader(receiverID, messageCounter, MESSAGE_ID_STRT, sizeof (HEABData), debug);
```
[the remark this answers is not in the document]
Comment (Angeleon) [reply iso22133.c:866]: hoppsan

File: src/grem.c [hunk 3]
```diff
@@ -171,13 +172,17 @@ enum ISOMessageReturnValue convertGREMoHostRepresentation(GREMType* GREMdata,
...
+ * \param messageCounter Message counter of GREM message
  * \param gremObjectData Struct containing relevant GREM data
  * \param gremDataBuffer Data buffer to which message is to be printed
  * \param bufferLength Available memory in data buffer
  * \param debug Flag for enabling debugging
  * \return Number of bytes written to buffer, or -1 in case of error
  */
-ssize_t encodeGREMMessage(const GeneralResponseMessageType* gremObjectData,
+ssize_t encodeGREMMessage(
+		const uint32_t receiverID,  const uint8_t messageCounter,
```
Comment (SebastianSvanland) [inline src/grem.c:184]:
```suggestion
		const uint32_t receiverID,  
		const uint8_t messageCounter,
```

File: src/dres.c [hunk 4]
```diff
@@ -4,16 +4,20 @@
...
+ * \param receiverID Receiver ID to be used for the message
+ * \param messageCounter Message counter to be used for the message
  * \param testObjectDiscoveryData Test object discovery data
  * \param dresDataBuffer Data buffer to which DRES is to be written
  * \param bufferLength Length of data buffer to which DRES is to be written
  * \param debug Flag for enabling debugging
  * \return Number of bytes written to buffer, or -1 in case of error
  */
 ssize_t encodeDRESMessage(
+	const uint32_t receiverID,  const uint8_t messageCounter,
```
Comment (SebastianSvanland) [inline src/dres.c:20]:
```suggestion
	const uint32_t receiverID,  
	const uint8_t messageCounter,
```

File: src/monr.c
Comment (SebastianSvanland) [inline src/monr.c:1]: I'm not familiar with the exact handling of every message. Why is objectID no longer needed? Does it serve a similar purpose to receiverID?

File: tests/grem.cpp [hunk 5]
```diff
@@ -17,8 +17,8 @@ class EncodeGREM : public ::testing::Test
 	}
 	void SetUp() override
 	{
-		memset(encodeBuffer, 0, sizeof(encodeBuffer));
-		auto res = encodeGREMMessage(&grem, encodeBuffer,
+		memset(&encodeBuffer, 0, sizeof(encodeBuffer));
+		auto res = encodeGREMMessage(0, 0x000000F0, &grem, encodeBuffer,
```
Comment (SebastianSvanland) [inline tests/grem.cpp:21]: Is this related to receiverID and messageCounter? Could they be given to descriptive variables instead that are used?

File: tests/header.cpp
Comment (SebastianSvanland) [inline tests/header.cpp:1]: Same as for GREM. What are these values?

File: tests/monr.cpp [hunk 6]
```diff
@@ -40,6 +40,7 @@ class EncodeMONR : public ::testing::Test {
 	{
 		memset(encodeBuffer, 0, sizeof(encodeBuffer));
 		auto res = encodeMONRMessage(
+			0, 0, // ReceiverID and MessageCounter
```
Comment (SebastianSvanland) [inline tests/monr.cpp:43]:
```suggestion
			0, // ReceiverID
			0, // MessageCounter
```

File: tests/ostm.cpp [hunk 7]
```diff
@@ -13,6 +13,7 @@ class EncodeOSTM : public ::testing::Test
 	{
 		memset(encodeBuffer, 0, sizeof(encodeBuffer));
 		auto res = encodeOSTMMessage(
+			0, 0, // ReceiverID and MessageCounter
```
Comment (SebastianSvanland) [inline tests/ostm.cpp:16]:
```suggestion
			0, // ReceiverID
			0, // MessageCounter
```

File: tests/traj.cpp [hunk 8]
```diff
@@ -11,6 +11,7 @@ class EncodeTRAJHeader : public ::testing::Test
 		char name[] = "some description";
 		memset(encodeBuffer, 0, sizeof(encodeBuffer));
 		auto res = encodeTRAJMessageHeader(
+			0, 0, //ReceiverID and MessageCounter
```
Comment (SebastianSvanland) [inline tests/traj.cpp:14]:
```suggestion
			0, // ReceiverID 
			0, // MessageCounter
```

File: tests/traj.cpp [hunk 9]
```diff
@@ -454,8 +454,10 @@ class EncodeTRAJFooter : public ::testing::Test
 	void SetUp() override
 	{
 		memset(encodeBuffer, 0, sizeof(encodeBuffer));
+		setTransmitterID(0x000000FF);
 		auto p = encodeBuffer;
 		auto offset = encodeTRAJMessageHeader(
+			0,0, //ReceiverID and MessageCounter
```
Comment (SebastianSvanland) [inline tests/traj.cpp:460]:
```suggestion
			0, // ReceiverID
			0, // MessageCounter
```

**Review by SebastianSvanland — COMMENTED** [review]
Looks simple enough for the most part. Have you also checked so that all the decoders are correct still? Or did this not change anything significant in that regard?

File: CMakeLists.txt [hunk 10]
```diff
@@ -18,6 +18,17 @@ FetchContent_Declare(googletest
 FetchContent_MakeAvailable(googletest)
```
Comment (sepast) [inline CMakeLists.txt:20]: There must be some instructions for how to enabling SWIG etc.
As default must the SWIG option be disabled.

File: include/header.h [hunk 11]
```diff
@@ -23,7 +23,7 @@ typedef struct {
 enum ISOMessageReturnValue decodeISOHeader(const char *MessageBuffer, const size_t length,
 											 HeaderType * HeaderData, const char debug);
 
-HeaderType buildISOHeader(enum ISOMessageID id, uint32_t messageLength, const char debug);
+HeaderType buildISOHeader(uint32_t receiverID, uint8_t messageCounter, enum ISOMessageID id, uint32_t messageLength, const char debug);
```
Comment (sepast) [inline include/header.h:27]: I would prefer this order of the parameters due to _enum ISOMessageID_ has priority:

HeaderType buildISOHeader(enum ISOMessageID id, uint32_t messageLength, uint32_t receiverID, uint8_t messageCounter, const char debug);

File: src/header.c [hunk 12]
```diff
@@ -6,21 +6,21 @@
...
 /*!
  * \brief buildISOHeader Constructs an ISO header based on the supplied message ID and content length
+ * \param receiverID ID of the receiver of the message for which the header is to be used
  * \param id Message ID of the message for which the header is to be used
  * \param messageLength Length of the message including header and footer
  * \param debug Flag for enabling debugging
  * \return A struct containing ISO header data
  */
-HeaderType buildISOHeader(enum ISOMessageID id, uint32_t messageLength, const char debug) {
+HeaderType buildISOHeader(uint32_t receiverID, uint8_t messageCounter, enum ISOMessageID id, uint32_t messageLength, const char debug) {
```
Comment (sepast) [inline src/header.c:17]: see comment about this row ...

**Review by sepast — COMMENTED** [review]
Nice work.
Does ATOS work with this PR?

File: include/header.h [hunk 13]
```diff
@@ -7,23 +7,10 @@ extern "C" {
 #include <stdint.h>
 #include "iso22133.h"
 
-#pragma pack(push,1)			// Ensure sizeof() is useable for (most) network byte lengths
-/*! ISO message header */
-typedef struct {
```
Comment (sepast) [inline include/header.h:12]: Where is this information?

File: CMakeLists.txt (hunk 10 above)
Comment (Angeleon) [reply CMakeLists.txt:20]: Will update README. Default for SWIG is off

File: include/header.h (hunk 13 above)
Comment (Angeleon) [reply include/header.h:12]: row 23

File: src/dres.c [hunk 14]
```diff
@@ -4,16 +4,19 @@
 #include <errno.h>
 #include "iso22133.h"
 #include "dres.h"
+#include "iohelpers.h"
```
Comment (sepast) [inline src/dres.c:7]: What's this?

File: src/header.c [hunk 15]
```diff
@@ -6,21 +6,21 @@
...
+HeaderType buildISOHeader(enum ISOMessageID id, const MessageHeaderType *input, uint32_t messageLength, const char debug) {
 	HeaderType header;
 
 	header.syncWord = ISO_SYNC_WORD;
-	header.transmitterID = getTransmitterID();
-	header.receiverID = 0; // TODO: Set receiver ID
-	header.messageCounter = 0;
+	header.transmitterID = input->transmitterID;
+	header.receiverID = input->receiverID;
+	header.messageCounter = input->messageCounter;
```
Comment (sepast) [inline src/header.c:23]: Maybe messageCounter shall be a static variable instead that is automatically incremented each time buildISOHeader(...) is called? To make it easier for the user.

File: src/ostm.c [hunk 16]
```diff
@@ -3,16 +3,19 @@
 #include <string.h>
 #include <endian.h>
 #include <errno.h>
+#include "iso22133.h"
```
Comment (sepast) [inline src/ostm.c:6]: Isn't iso22133.h already included?

File: src/traj.c [hunk 17]
```diff
@@ -1,5 +1,6 @@
 #include "traj.h"
 #include "iohelpers.h"
+#include "iso22133.h"
```
Comment (sepast) [inline src/traj.c:3]: Already included?

**Review by sepast — COMMENTED** [review]
Looks good.
Some comments.
We need to review the ATOS branch with this branch before approval.

File: src/dres.c (hunk 14 above)
Comment (Angeleon) [reply src/dres.c:7]: No clue. Legacy code that was needed in this file

File: src/header.c (hunk 15 above)
Comment (Angeleon) [reply src/header.c:23]: Messagecounter must be handled by the modules using this code, as there is no way of knowing which counter is to be used

File: src/traj.c (hunk 17 above)
Comment (Angeleon) [reply src/traj.c:3]: no, but needed

File: tests/header.cpp
Comment (Angeleon) [reply tests/header.cpp:1]: fixed

File: include/header.h [hunk 18]
```diff
@@ -17,13 +17,13 @@ typedef struct {
 	uint32_t receiverID;
 	uint8_t messageCounter;
 	uint16_t messageID;
-} HeaderType;
+} HeaderType;;
```
Comment (victorjarlow) [inline include/header.h:20]:
```suggestion
} HeaderType;
```

File: iso22133.i [hunk 19]
```diff
@@ -4,33 +4,42 @@
...
+#include "dres.h"
+#include "footer.h"
+#include "grem.h"
+#include "header.h"
+#include "iohelpers.h"
+#include "monr.h"
+#include "osem.h"
+#include "ostm.h"
+#include "timeconversions.h"
+// #include "traj.h"
```
Comment (victorjarlow) [inline iso22133.i:18]:
```suggestion
```

File: include/defines.h [hunk 20]
```diff
@@ -1,4 +1,7 @@
 #pragma once
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Comment (victorjarlow) [inline include/defines.h:4]: What is the purpose of this?

File: src/dreq.c [hunk 21]
```diff
@@ -4,15 +4,17 @@
 #include <errno.h>
 #include "iso22133.h"
 #include "dreq.h"
+#include "iohelpers.h"
```
Comment (victorjarlow) [inline src/dreq.c:7]: Is this needed?

File: src/dres.c (hunk 14 above)
Comment (victorjarlow) [inline src/dres.c:7]:
```suggestion
```

File: src/grem.c [hunk 22]
```diff
@@ -5,6 +5,7 @@
 #include "grem.h"
 #include "iohelpers.h"
 
+
```
Comment (victorjarlow) [inline src/grem.c:8]:
```suggestion
```

**Review by victorjarlow — CHANGES_REQUESTED** [review]
Is swig necessary for the changes with receiver ID and header of the ISO protocol? If not, split into another PR that deals exclusively with swig-parts

File: iso22133.i [hunk 23]
```diff
@@ -4,33 +4,42 @@
...
-extern ssize_t decodeOSEMMessage(ObjectSettingsType *objectSettingsData, const char * osemDataBuffer, const size_t bufferLength, uint32_t *senderID, const char debug);
-%}
 
 
 #%javaconst(1);
 
 typedef double double_t;
 typedef long int ssize_t;
 
-struct timeval {
```
Comment (vincebel8) [inline iso22133.i:26]: Why was this removed?

File: src/dreq.c (hunk 21 above)
Comment (Angeleon) [inline src/dreq.c:7]:
```suggestion
```

File: iso22133.i (hunk 23 above)
Comment (Angeleon) [reply iso22133.i:26]: it was an superfluous define, which caused problems when building SWIG. See positioning.i

**Review by Robert108 — APPROVED** [review]
